### PR TITLE
refactor: moved inversion functionality to inverses.py

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Refactoring (no functional changes)
- Tests

### Description

Adds helper functions to `RegularisedInverseApproximator` to map approximations across first axis.
Moves `invert_regularised_array` from `util.py`  from `inverses.py`  as `LeastSquareApproximator`.

Increased default `oversampling_parameter` after more testing with CMMD solver.

### How Has This Been Tested?

Moved `invert_regularised_array` tests from `test_util.py` to  `test_inverses.py` .
Added additional approximation map tests.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/approximation.py [hunk 1]
```diff
@@ -338,11 +338,75 @@ def approximate(
...
+            identity,
+            rcond=self.rcond,
+        )[0]
+
 
 def randomised_eigendecomposition(
     random_key: KeyArrayLike,
     array: Array,
-    oversampling_parameter: int = 10,
+    oversampling_parameter: int = 25,
```
Comment (tc85324) [inline coreax/approximation.py:409]: This default seems to be arbitrary, is there a well-reasoned default that can be computed by making this optional (default None)?

File: coreax/approximation.py [hunk 2]
```diff
@@ -338,11 +338,75 @@ def approximate(
...
+                raise ValueError("'rcond' must be non-negative, except for value of -1")
+
+    @override
+    def approximate(
+        self,
+        kernel_gramian: Array,
+        regularisation_parameter: float,
+        identity: Array,
+    ) -> Array:
+        if kernel_gramian.shape != identity.shape:
```
Comment (tc85324) [inline coreax/approximation.py:396]: The JAX error raised here is essentiallly the same, "Leading dimensions of input arrays must match". If you want to name these arrays, then you should utilise a try-except here, instead of an explicit if statement. This try-except pattern is consitent with the rest of the codebase and is generally preferd for two reasons. 1. It allows us to be more permisive towards user input, only raising when the underlying problem is triggered. 2. It yields a more informative stack-trace; the current approach will tell the user the leading dimensions need to match, but won't make clear why they need to match; a try-except approach would raise from the underlying error raised in `lstsq`, making it more obvious why this dimension mismatch is a problem.

**Review by tc85324 — COMMENTED** [review]
My primary concerns here are that the `approximate.py` module is for approximating (overloading) the methods of `coreax.kernel.Kernel`s. My secondary concern is that the inversion methods are (I think) unnecessarily (conceptually) coupled to the existance of a kernel or one of its matrices (E.G. the Gramian). Can these methods be generalised to the inversion of arbitraray matrices (subject to certain conditions)? If not, then I think these should be called `BlockInverseApproximators` or something similar, and moved into their own module `inverses.py`. 

If the kernel coupling must exist, then maybe we want to add a `compute_inverse` method to kernel and have these approximators overload it? This way we are still defining `ApproximateKernels`, and we can compose multiple approximations.

File: coreax/approximation.py (hunk 1 above)
Comment (db091756) [reply coreax/approximation.py:410]: From #637:

From the Halko paper, where p is the oversampling_parameter

_Just how many extra samples do we need? Remarkably, for certain types of
random sampling schemes, the failure probability decreases superexponentially with
the oversampling parameter p. As a practical matter, setting p = 5 or p = 10
often gives superb results. This observation is one of the principal facts supporting
the randomized approach to numerical linear algebra._

Also see:

_The number of columns ℓ that the algorithm needs to reach this threshold is usually slightly larger than the minimal rank k of the smallest basis that verifies (4.2; here 4.2. is the error the authors consider for the matrix that has a similar range to the target array). We refer to this discrepancy p = ℓ − k as the oversampling parameter. The size of the oversampling parameter depends on several factors:_

_1. The matrix dimensions. Very large matrices may require more oversampling.
2. The singular spectrum. The more rapid the decay of the singular values, the less oversampling is needed. In the extreme case that the matrix has exact rank k, it is not necessary to oversample._

_Some experimentation led me to use larger values of p for the tests as the kernel matrix I used does not always have a rapidly decaying singular spectrum. The oversampling_parameter  should also not be any larger than the size of the matrix as then you are gaining nothing by this approach.

It could be worth implementing Algorithm 4.2 from the paper as this avoids leaving the choice of oversampling_parameter to the user._

On the power_iterations front Halko has this to say:

_Unfortunately, the simplest version of this scheme (referring to approximate eigen-decomposition) is inadequate in many
applications because the singular spectrum of the input matrix may decay slowly. To
address this difficulty, we incorporate q steps of a power iteration, where q = 1 or
q = 2 usually suffices in practice._

There is an element of diminishing returns, especially with power_iterations, it is one of the more costly parts of the algorithm and once you have reduced the error to machine precision it is pointless to keep beating the horse. I have also found that you can actually make things worse with too many power_iterations, probably due to floating-point errors.

**Comment by db091756** [comment]
> My primary concerns here are that the `approximate.py` module is for approximating (overloading) the methods of `coreax.kernel.Kernel`s. My secondary concern is that the inversion methods are (I think) unnecessarily (conceptually) coupled to the existance of a kernel or one of its matrices (E.G. the Gramian). Can these methods be generalised to the inversion of arbitraray matrices (subject to certain conditions)? If not, then I think these should be called `BlockInverseApproximators` or something similar, and moved into their own module `inverses.py`.
> 
> If the kernel coupling must exist, then maybe we want to add a `compute_inverse` method to kernel and have these approximators overload it? This way we are still defining `ApproximateKernels`, and we can compose multiple approximations.

The functions do work independently of the kernels, I would be happy to uncouple them and move them to their own file.

File: coreax/approximation.py (hunk 1 above)
Comment (tc85324) [reply coreax/approximation.py:397]: Ok that makes sense to me, thanks! It might be useful to point the user to these in the docstrings, or explain the default. On a side note, inside the docstrings, the parameters should be surrounded by backticks instead of single quotes.

**Review by tc85324 — CHANGES_REQUESTED** [review]
Appologies for the dealy in getting back to this.

I'd like a few clarifications on mathematical terminology, mostly to check my own understading. Also, am I right to think all these methods be generalized to the solution of general (non-)linear systems?

Note there are also some pre-commit failures (pylint no-member)

File: coreax/inverses.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+.. math::
+    A^{-1} := \begin{bmatrix}
+        B^{-1} & 0 & \dots & 0 \\
+        0 & 0 & \dots & 0 \\
+        \vdots & \ddots & \dots & \vdots \\
+        0 & 0 & \dots & 0
+    \end{bmatrix}.
+
+This functionality allows iterative coreset algorithms which require inverting growing
+arrays to have fully static array shapes and thus be JIT-compilable in JAX.
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:52]: I meant it to be `compilable`, like is possible to compile.

**Comment by db091756** [comment]
> Appologies for the dealy in getting back to this.
> 
> I'd like a few clarifications on mathematical terminology, mostly to check my own understading. Also, am I right to think all these methods be generalized to the solution of general (non-)linear systems?
> 
> Note there are also some pre-commit failures (pylint no-member)

Now worries and thanks for the review! The pre-commit failures come from parts of the codebase that this pull doesn't touch and aren't being flagged locally so not sure what is happening. They stay when I re-run it also.

File: .cspell/custom_misc.txt [hunk 4]
```diff
@@ -16,6 +16,8 @@ elementwise
 forall
 GCHQ
 Gramian
+gramians
+Hermitan
```
[the remark this answers is not in the document]
Comment (db091756) [reply .cspell/custom_misc.txt:20]: Yep, good spot

File: coreax/inverses.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+The most efficient way to compute these "inverses" in JAX requires the `identity` array
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:54]: Good point, will remove.

File: coreax/inverses.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+    @abstractmethod
+    def approximate(
+        self,
+        array: Array,
+        regularisation_parameter: float,
+        identity: Array,
+    ) -> Array:
+        r"""
+        Approximate inverse of regularised array.
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:95]: On the regularisation point, what we are doing here is solving (or approximating) a regularised least squares problem (https://en.wikipedia.org/wiki/Regularized_least_squares), it is just that in this case the solution is a approximately the inverse of our target array. So I could rewrite the docstring to reflect that fact more clearly.

On the `approximate` front, the code returns an array that exactly solves the regularised least-square problem above, in the sense that it has the lowest L2 error like you say, but this doesn't mean that you get an exact inverse out at the end. So we'd have to pin down what exactly we mean by approximate, I meant it in the sense that the final inverse is an approximation. Because our target arrays are going to be kernel matrices here which are generally ill-conditioned you can obviously get very "approximate" solutions, if you are not careful.

File: coreax/inverses.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        :param identity: Block identity matrix
+        :return: Approximation of the kernel matrix inverses
+        """
+        _map_approximate = jax.vmap(self.approximate, in_axes=(0, None, None))
+        return _map_approximate(arrays, regularisation_parameter, identity)
+
+
+class LeastSquareApproximator(RegularisedInverseApproximator):
+    """
+    Approximate inverse of regularised array by solving a least-squares problem.
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:131]: Yep! That seems like a correct understanding to me

File: coreax/inverses.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Classes and associated functionality to approximate regularised inverses of matrices.
+
+Given a matrix :math:`A\in\mathbb{R}^{n\times n}` and some regularisation parameter
+:math:`\lambda\in\mathbb{R}_{\ge 0}`, the regularised inverse of :math:`A` is
+:math:`(A + \lambda I_n)^{-1}` where :math:`I_n` is the :math:`n\times n` identity
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:20]: See (https://en.wikipedia.org/wiki/Regularized_least_squares), you can see that Tikhonov regularisation is a specific example using the L2-norm like you say.

File: coreax/inverses.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+    """
+
+    oversampling_parameter: int = 25
+    power_iterations: int = 1
+    rcond: Optional[float] = None
+
+    def __check_init__(self):
+        """Validate rcond input."""
+        if self.rcond is not None:
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:264]: In this case we do need to check this ourselves because we are not calling any in-built functions that use and check values of rcond. We use it ourselves directly.

File: tests/unit/test_inverses.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+    def approximator(self) -> _RegularisedInverseApproximator:
+        """Abstract pytest fixture which initialises an inverse approximator."""
+
+    @abstractmethod
+    def problem(self) -> _Problem:
+        """Abstract pytest fixture which returns a problem for inverse approximation."""
+
+    def test_approximation_accuracy(
+        self, problem: _Problem, approximator: _RegularisedInverseApproximator
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (db091756) [reply tests/unit/test_inverses.py:69]: Removed them

File: coreax/inverses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/inverses.py:44]: Ok, I see. Because `jax.jit` transforms a function (it acts on a python function) I tend to prefer the phrase 'JIT compatible' (compatible can be freely interchanged with transformable). That said, the main reason I would prefer not to use 'compilable' here is because it has no entry in the OED and can be easily misread as compliable (which I did).

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py:44]: Makes sense, changed

**Review by tc85324 — CHANGES_REQUESTED** [review]
A small typo and continued discussion on terminology. I've also added the issue #658,  which, while outside the scope of this PR, may be worth future consideration.

File: coreax/inverses.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/inverses.py]: Thanks for looking into this more thoroughly, I will try incorporate the gist of what you have found into the docstrings and argue the case for why we need an approximation.

**Review by tc85324 — APPROVED** [review]
I'm happy to approve this! I think we could spend more time trying to refine the precise wording of all the terms, but it would probably be little more than an academic exercise. The discussion in the prior reviews is enough to satisfy my concerns!
